File: ios/StackSpeak/Models/ProgressSnapshot.swift
import Foundation

/// The platform-neutral record that syncs across devices and platforms.
///
/// This is the **cross-platform contract**: iOS and (later) Android must
/// serialize the *identical* JSON shape so a user's progress is portable
/// between an iPhone, an iPad, and an Android device. It is stored as one
/// compact JSONB blob per user (see `supabase/migrations`).
///
/// Deliberately excluded: entitlement (`isPro`/`proExpiryDate`/`isLifetimePro`
/// — each device derives Pro from its own store), transient daily counters,
/// and device-local preferences (notifications, theme). Only learning progress
/// travels.
///
/// `schemaVersion` lets future shape changes migrate safely; `updatedAt` drives
/// last-write/merge decisions during sync.
struct ProgressSnapshot: Codable, Equatable {
    static let currentSchemaVersion = 1

    var schemaVersion: Int
    var updatedAt: Date

    // Core progression
    var level: Int
    var currentStreak: Int
    var longestStreak: Int
    var lastCompletedDate: Date?
    var didCompleteOnboarding: Bool

    // Word-id sets (stable mnemonic-derived UUIDs, safe across platforms)
    var practicedWordIds: [String]
    var masteredWordIds: [String]
    var bookmarkedWordIds: [String]
    var wordsWithTwoCorrectIds: [String]
    var wordsCreditedForLevelIds: [String]

    // Rotation state — kept so the daily queue stays consistent across devices
    var selectedStacks: [String]
    var shuffleSeed: String
    var wordQueueCursor: Int

    // Sub-records
    var reviewStates: [ReviewStateDTO]
    var assessmentResults: [AssessmentResultDTO]
    var practicedSentences: [PracticedSentenceDTO]
    var bookProgress: [BookProgressDTO]

    struct ReviewStateDTO: Codable, Equatable {
        var wordId: String
        var easinessFactor: Double
        var interval: Int
        var repetitions: Int
        var dueDate: Date
        var lastReviewedAt: Date?
    }

    struct AssessmentResultDTO: Codable, Equatable {
        var id: String
        var wordId: String
        var attemptedAt: Date
        var isCorrect: Bool
        var selectedAnswer: String
        var correctAnswer: String
    }

    struct PracticedSentenceDTO: Codable, Equatable {
        var wordId: String
        var sentence: String
        var createdAt: Date
        var inputMethod: String
    }

    struct BookProgressDTO: Codable, Equatable {
        var bookId: String
        var lastOpenedAt: Date
        var currentChapterId: String?
        var currentCardId: String?
        var completedCardIds: [String]
        var lastReadingDayString: String
        var currentStreakDays: Int
        var longestStreakDays: Int
    }
}

File: android/app/src/main/java/com/stackspeak/domain/ProgressSnapshot.kt
package com.stackspeak.domain

import kotlinx.serialization.Serializable
import java.time.Instant

/**
 * Platform-neutral sync record — the cross-platform contract. Must serialize to
 * the *identical* JSON shape as iOS `ProgressSnapshot.swift` so a user's progress
 * is portable iPhone ↔ iPad ↔ Android. One JSONB blob per user.
 *
 * Parity rules (see shared/test-fixtures/README.md): camelCase keys = property
 * names; dates ISO-8601 with no fractional seconds (Z); nil optionals are omitted
 * keys, not null; word-id strings are UPPERCASE. Entitlement, daily counters, and
 * device prefs are deliberately NOT here.
 */
@Serializable
data class ProgressSnapshot(
    val schemaVersion: Int,
    @Serializable(InstantIso8601Serializer::class) val updatedAt: Instant,

    val level: Int,
    val currentStreak: Int,
    val longestStreak: Int,
    @Serializable(InstantIso8601Serializer::class) val lastCompletedDate: Instant? = null,
    val didCompleteOnboarding: Boolean,

    val practicedWordIds: List<String>,
    val masteredWordIds: List<String>,
    val bookmarkedWordIds: List<String>,
    val wordsWithTwoCorrectIds: List<String>,
    val wordsCreditedForLevelIds: List<String>,

    val selectedStacks: List<String>,
    val shuffleSeed: String,
    val wordQueueCursor: Int,

    // v2 (2026-06). Optional/defaulted so pre-v2 rows still decode; a missing
    // value means the default 5 on apply.
    val dailyWordGoal: Int? = null,

    val reviewStates: List<ReviewStateDTO>,
    val assessmentResults: List<AssessmentResultDTO>,
    val practicedSentences: List<PracticedSentenceDTO>,
    val bookProgress: List<BookProgressDTO>,
) {
    companion object {
        const val CURRENT_SCHEMA_VERSION = 2
    }

    @Serializable
    data class ReviewStateDTO(
        val wordId: String,
        val easinessFactor: Double,
        val interval: Int,
        val repetitions: Int,
        @Serializable(InstantIso8601Serializer::class) val dueDate: Instant,
        @Serializable(InstantIso8601Serializer::class) val lastReviewedAt: Instant? = null,
    )

    @Serializable
    data class AssessmentResultDTO(
        val id: String,
        val wordId: String,
        @Serializable(InstantIso8601Serializer::class) val attemptedAt: Instant,
        val isCorrect: Boolean,
        val selectedAnswer: String,
        val correctAnswer: String,
    )

    @Serializable
    data class PracticedSentenceDTO(
        val wordId: String,
        val sentence: String,
        @Serializable(InstantIso8601Serializer::class) val createdAt: Instant,
        val inputMethod: String,
    )

    @Serializable
    data class BookProgressDTO(
        val bookId: String,
        @Serializable(InstantIso8601Serializer::class) val lastOpenedAt: Instant,
        val currentChapterId: String? = null,
        val currentCardId: String? = null,
        val completedCardIds: List<String>,
        val lastReadingDayString: String,
        val currentStreakDays: Int,
        val longestStreakDays: Int,
    )
}
